### ISSUE TITLE
write.wap.enabled / spark.wap.branch / spark.wap.id behavior isn't really documented

### ISSUE THREAD

**Issue by wimlewis-amazon** [apache/iceberg#11528]
### Feature Request / Improvement

It's unclear from reading the doc what the effect of setting `write.wap.enabled` actually is. I can make some assumptions based on the write-audit-publish example, but those assumptions are almost certainly incomplete (for example, there are edge cases and failure conditions that I would need to understand before applying this in production, but the example only shows the "happy path".). As a user of the Iceberg libraries, I would like the documentation to be more explicit about what changes if I set either of these configuration options.

### Query engine

Spark

### Willingness to contribute

- [ ] I can contribute this improvement/feature independently
- [X] I would be willing to contribute this improvement/feature with guidance from the Iceberg community
- [ ] I cannot contribute this improvement/feature at this time

**Comment by rodrigc** [comment]
I also did not fully understand `write.wap.enabled`.
Looking at the code, it seems like an additional configuration knob
that can turn off WAP even if `spark.wap.branch` is set: 

https://github.com/apache/iceberg/blob/17bda20b2901ab76adac1339dccd79656b913d33/spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/SparkReadConf.java#L123
https://github.com/apache/iceberg/blob/17bda20b2901ab76adac1339dccd79656b913d33/core/src/main/java/org/apache/iceberg/TableProperties.java#L310

**Comment by wimlewis-amazon** [comment]
Hmm, from those code snippets, it looks like `enabled` is a _table_ property which enables the behavior, and `branch` is a _spark session_ property that indicates which branch to _read_ (and alternatively disables WAP if it's unset). That's a little weird (shouldn't at least one of those be a _write options_ property? And it seems that WAP affects the behavior of table read as well, which isn't what I would naively expect, although from the PR discussions I can kind of see the idea.)

Digging through git-blame, I found a bunch of discussion, in PRs #6651, #6965, which references these behaviors but it's still not clear what the final intended behavior is — what I should expect as a user of the spark api.

**Comment by rodrigc** [comment]
@wimlewis-amazon my understanding is that:
1.  `write.wap.enabled` is a table property, and must be set on the table in order to do WAP
2.  `spark.wap.branch` is a SQL conf property that specifies the name of the branch

Looking at the code, it looks like you need to have the table property *and* the SQL conf property set in order
to do WAP.
This lines up with the example in the docs here: https://iceberg.apache.org/docs/latest/branching/#audit-branch

Originally I looked at the example in the docs, and didn't really understand why there was a table property
and an SQL conf property for WAP.  Looking at the code snippets above improved my understanding.
But I admit, I find this a bit confusing as an end-user.

**Comment by wimlewis-amazon** [comment]
This is still a pretty relevant issue.

**Comment by Dima1224** [comment]
I agree the documentation is lacking in this regard. Specifically, I am trying to understand:

- Can I write to a table which is `write.wap.enabled=true` in the standard ways and expect standard results?
- What exactly is the effect of setting `write.wap.enabled=true` and why is this flag necessary?

I found this thread helpful, but it would be nice to see this info in the docs.

**Comment by gphilipp** [comment]
I'm a bit confused too. I've set `spark.wap.id` to some value. Now if I set `spark.wap.branch` as well, it fails with:

```
org.apache.hive.service.cli.HiveSQLException: Error running query: org.apache.iceberg.exceptions.ValidationException: Cannot set both WAP ID and branch
```

It seems they’re mutually exclusive. What’s the intended difference between spark.wap.id and spark.wap.branch? When should each be used?
